Excel_Adapter prototype re-implemented

pawelbaran opened the pull request:
### Issues addressed by this PR
<!-- Add reference(s) to issue(s) solved by this PR. Please use keyword Fixes/Closes as per https://help.github.com/articles/closing-issues-using-keywords/ -->

Closes #221

<!-- Add short description of what has been fixed -->


### Test files
<!-- Link to test files to validate the proposed changes -->
On [SharePoint](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?RootFolder=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F02%5FPull%20Request%2FBHoM%2FExcel%5FToolkit%2F%23320%2DAdapter&FolderCTID=0x0120008122C8891F89054B8ACED0196C70DFC4).


### Changelog
<!-- Text to go into changelog if applicable -->
<!-- Please see https://github.com/BHoM/documentation/wiki/changelog for guidelines -->


### Additional comments
<!-- As required -->
This is another iteration of https://github.com/BHoM/Excel_Toolkit/pull/241.

The general concept is that the only object type we can push or pull is `BH.oM.Data.Collections.Table` - each of these represents one sheet in an Excel workbook. So if we `Push` a few tables, a few sheets will be created/updated, on `Pull` the output object count should be equal to the number of extracted sheets. I believe this rationale is simple yet powerful and brings exactly what we need:
- ability to push tabular data
- ability to pull tabular data

No more than that. If one wants to work with more complex objects - Sockets etc. is the way to go.

Looking forward to hearing others' opinions!

pawelbaran commented:
@BHoMBot check compliance
@BHoMBot check null-handling

pawelbaran commented:
@BHoMBot fix copyright headers ref. 4225699803

pawelbaran commented:
@BHoMBot fix project file ref. 4225698002

pawelbaran commented:
@BHoMBot check compliance

pawelbaran commented:
@BHoMBot check null-handling

pawelbaran commented:
@BHoMBot check compliance
@BHoMBot check core
@BHoMBot check installer
@BHoMBot check serialisation
@BHoMBot check versioning

rolyhudson commented on Excel_Adapter/AdapterActions/Push.cs:
```diff
@@ -0,0 +1,190 @@
+                success &= Update(workbook, table, config);
+            }
+
+            // Try to update the workbook mateinformation and then save it.
```
```suggestion
            // Try to update the workbook properties and then save it.
```

pawelbaran replied:
It was meant to be metainformation 😆 But changed to properties (not sure what is the right term for all that document info).

rolyhudson commented on Excel_Adapter/CRUD/Read/Read.cs:
```diff
@@ -0,0 +1,159 @@
+            }
+            catch
+            {
+
```
If the catch is needed shouldn't we have an error message here?

pawelbaran replied:
No need: this catch means that `workbook` keeps on being `null` and is caught by a null check (and correspondent error) in lines 54-58.

FraserGreenroyd replied:
I would add a code comment to this effect @pawelbaran to avoid confusion for future maintainers

rolyhudson reviewed (changes requested):
An inspired PR 🤣 . Been through the test script and the code. Two minor comments but otherwise great work. I will be able to put this to work immediately.

On a side note: this create table method:
![image](https://user-images.githubusercontent.com/6618854/142253262-28940bbd-f446-46ba-9ac2-57ac791cf0a5.png)
Is very confusing.
Which appears to be hard coded to never produce a table with more than three columns.
I've also tested with other tables generated from BHoM objects and works great.

pawelbaran commented:
Thanks @rolyhudson for the review, I have introduce the requested change.

> On a side note: this create table method: ![image](https://user-images.githubusercontent.com/6618854/142253262-28940bbd-f446-46ba-9ac2-57ac791cf0a5.png) Is very confusing. Which appears to be hard coded to never produce a table with more than three columns.

Agreed, it confused me a lot too. However, I had hard times producing more complex tables in GH, so assumed that this should be enough for the sake of testing (if it works for 3x10 table, I assume it should also work for 100x100). Happy that you managed to test it on more complex datasets, big thanks for that! 👍

FraserGreenroyd commented on Excel_Adapter/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,58 @@
+// You can specify all the values or you can default the Build and Revision Numbers
+// by using the '*' as shown below:
+// [assembly: AssemblyVersion("1.0.*")]
+[assembly: AssemblyVersion("1.0.0.0")]
+[assembly: AssemblyFileVersion("1.0.0.0")]
```
```suggestion
[assembly: AssemblyVersion("5.0.0.0")]
[assembly: AssemblyFileVersion("5.0.0.0")]
```

FraserGreenroyd commented on Excel_Adapter/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,58 @@
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("")]
+[assembly: AssemblyProduct("Excel_Adapter")]
+[assembly: AssemblyCopyright("Copyright ©  2020")]
```
This needs updating to match the other assembly infos

FraserGreenroyd commented on Excel_oM/ClosedXML/CellContents.cs:
```diff
@@ -0,0 +1,53 @@
+
+namespace BH.oM.Adapters.Excel
+{
+    public class CellContents : BHoMObject
```
I think this should be `CellContent` - one object == one cell?

FraserGreenroyd replied:
Class properties also need some descriptions, it's not immediately obvious to me what `FormulaR1C1` will do if set.

pawelbaran replied:
> it's not immediately obvious to me what `FormulaR1C1` will do if set.

Welcome to the club, sir 🤣 But fair point, will figure it out and update.

FraserGreenroyd replied:
🤣

rolyhudson replied:
Definitely should be `CellContents` as content is an uncountable noun eg content of a speech.

alelom replied:
Following the general convention, I would have this method as a `Convert` rather than a `Create`.

FraserGreenroyd commented on Excel_oM/ClosedXML/CellContents.cs:
```diff
@@ -0,0 +1,53 @@
+
+        public virtual object Value { get; set; } = null;
+
+        public virtual string Address { get; set; } = "";
```
Is this meant to be Cell Address, e.g. B4, C5, etc.?

Is it worth splitting into Row and Column as separate values?

pawelbaran replied:
Yeah, you are right about the cell address. I had a bit of thinking about the cell address/range format, and decided to follow the standard Excel format for a few reasons:
- To me, an Excel dummy, C5 is more readable than row 5, column 3.
- The above get particularly relevant when some rows/columns are hidden, and I have no clue what is the number any more, plus can't even count it. Imagine a spreadsheet with anything between F and AB being hidden - visually, AC column is now under index 6, how could I know its actual number?
- Actually, I have no clue if the cells/rows cannot be re-labelled with some Excel black magic, which would make the concept of row/column indices even more error-prone.
- The format used here is consistent with `ClosedXML`, which I assumed knows better 🙈 

Taken all above, I leaned towards simplicity and (my perception of) user-friendliness. Similar refers [this comment](https://github.com/BHoM/Excel_Toolkit/pull/320#discussion_r755975940). However, I do not have a very strong opinion on the topic, so happy to continue the discussion if you believe your alternative is better @FraserGreenroyd 👍

pawelbaran replied:
Looping @rolyhudson @IsakNaslundBh @alelom in - let's see what others think.

FraserGreenroyd replied:
So I was thinking it would still be `Column C` and `Row 5` to denote C5, rather than going to column numbers, but just splitting it into `ColumnAddress` and `RowAddress` to make some parametric options easier (duplicating objects for the same row number but different columns, etc.).

I would agree with not trying to convert `6` into `F` in the code given the possibility of hidden columsn.

rolyhudson replied:
I originally saw this object as a simplified wrapper of the  ​[closedxml `IXLCell` ](https://github.com/ClosedXML/ClosedXML/blob/78150efbbd4a36d65e95ef3c793f12feb12c1a9c/ClosedXML/Excel/Cells/IXLCell.cs)
With its intended use for accessing data beyond just the value associated with the cell.
`ValuesRequest` just reads the values
ClosedXML use of Address is in the format "C5". And I think we should keep the same. 
With ClosedXML cells can be accessed either by Address or using row and column
I think adding properties
int Row
int Cell
So we have all options.

pawelbaran replied:
I kept this and [this](https://github.com/BHoM/Excel_Toolkit/pull/320#discussion_r755975940) comment unresolved, but so far I would rather stick to the current solution (to large extent out of laziness 😃).

An alternative for the simple would be:
- `CellAddress` class with `string Column` and `int Row` properties
- `CellRange` class with `CellAddress From` and `CellAddress To` properties

Implementation should be really simple, (no more than string split or concatenation on call to ClosedXML), so no prob in getting this in - do you see any issues with that @rolyhudson? Let's maybe also wait for @alelom, who is also meant to have a look at this PR today.

alelom replied:
I think it's worth supporting both `string` and `int` inputs for the `Column` indexing. It's a classical issue when dealing with Excel that has been resolved many times.

I don't have an opinion on having a `CellAddress` class or not. Regardless of that, given the fields for the column and row index, I would have them as:
- `int RowIndex`
- `object ColumnIndex`

the `ColumnIndex` can then be converted to an integer or a string as appropriate by means of [a function like this](https://stackoverflow.com/a/182924/3873799).

This way, users could use progressive numbers for columns and/or the "classical lettered approach" from Excel.

FraserGreenroyd commented on Excel_oM/ClosedXML/WorkbookProperties.cs:
```diff
@@ -0,0 +1,57 @@
+        /**** Properties                                ****/
+        /***************************************************/
+
+        public virtual string Author { get; set; } = "author";
```
Should these be the defaults? Would it not be best to default to empty strings for these?

pawelbaran replied:
Yeah this was taken from @rolyhudson's prototype, I was wondering about it too, but finally left it as is. However, thinking about it now, I think it would be more consistent to keep the values blank as default.

FraserGreenroyd commented on Excel_oM/Requests/CellsRequest.cs:
```diff
@@ -0,0 +1,45 @@
+namespace BH.oM.Adapters.Excel
+{
+    [Description("IRequest that pulls cells and their metadata from the excel file.")]
+    public class CellsRequest : IRequest
```
Might it be better to call this `CellRangeRequest` - as `CellRange` is closer to the language used in Excel so might be more intuitive for users?

pawelbaran replied:
Hmm not sure, the range input is actually optional, so I would be afraid of being too indicative about the range input. Maybe others would have another opinion on that @rolyhudson @IsakNaslundBh @alelom?

rolyhudson replied:
I think it should be `CellContentsRequest` as it will return a collection of `CellContents` (formulas, datatype, hyperlink etc) not just the value. Cell values are accessed by 'ValuesRequest`

pawelbaran replied:
I changed the name to `CellContentsRequest`, as suggested by @rolyhudson, which makes most sense to me. Happy to refine further if others disagree 👍

FraserGreenroyd commented on Excel_oM/Requests/ValuesRequest.cs:
```diff
@@ -0,0 +1,45 @@
+        public virtual string Worksheet { get; set; } = "";
+
+        [Description("Range to read, in standard Excel format (e.g. A1:B6). If left empty, entire extent of the spreadsheet starting from A1 is used.")]
+        public virtual string Range { get; set; } = "";
```
Is it worth implementing a `Range` object of our own to avoid users having to use the Excel way of something : something (with the : being the contentious bit for me)?

FraserGreenroyd reviewed (dismissed):
Some comments for discussion, some descriptions needed (which is the cause for request of changes), otherwise this is looking good as an initial prototype.

alelom commented on Excel_Adapter/AdapterActions/Push.cs:
```diff
@@ -0,0 +1,190 @@
+                config = new ExcelPushConfig();
+            }
+
+            // Preprocess the objects to push.
+            IEnumerable<IBHoMObject> objectsToPush = ProcessObjectsForPush(objects, actionConfig);
+            if (!objectsToPush.Any())
+                return new List<object>();
```
Not needed here. ProcessObjectsForPush is only needed when we support a variety of objects and the full CRUD (update). In this case we are already imposing a constraint on the input objects being `Table`.

alelom commented on Excel_Adapter/AdapterActions/Push.cs:
```diff
@@ -0,0 +1,190 @@
+                }
+            }
+
+            // Make sure that only objects to be pushed are Tables.
+            List<Type> objectTypes = objects.Select(x => x.GetType()).Distinct().ToList();
+            if (objectTypes.Count != 1)
+            {
+                string message = "The Excel adapter only allows to push objects of a single type to a table."
+                    + "\nRight now you are providing objects of the following types: "
+                    + objectTypes.Select(x => x.ToString()).Aggregate((a, b) => a + ", " + b);
+                Engine.Reflection.Compute.RecordError(message);
+                return new List<object>();
+            }
```
I would move this up, before attempting to access the file.

alelom commented on Excel_Adapter/AdapterActions/Push.cs:
```diff
@@ -0,0 +1,190 @@
+                }
+            }
+
+            // Make sure that only objects to be pushed are Tables.
+            List<Type> objectTypes = objects.Select(x => x.GetType()).Distinct().ToList();
+            if (objectTypes.Count != 1)
+            {
+                string message = "The Excel adapter only allows to push objects of a single type to a table."
+                    + "\nRight now you are providing objects of the following types: "
+                    + objectTypes.Select(x => x.ToString()).Aggregate((a, b) => a + ", " + b);
+                Engine.Reflection.Compute.RecordError(message);
+                return new List<object>();
+            }
+
+            Type type = objectTypes[0];
+            if (type != typeof(Table))
+            {
+                BH.Engine.Reflection.Compute.RecordError($"Push failed: Excel Adapter can push only one objects of type {nameof(Table)}.");
+                return new List<object>();
+            }
+
+            // Check if all tables have distinct, non-empty names.
+            List<Table> tables = objects.Cast<Table>().ToList();
+            if (tables.Any(x => string.IsNullOrWhiteSpace(x.Name)))
+            {
+                BH.Engine.Reflection.Compute.RecordError("Push aborted: all tables need to have non-empty name.");
+                return new List<object>();
+            }
+
+            List<string> duplicateNames = tables.GroupBy(x => x.Name.ToLower()).Where(x => x.Count() != 1).Select(x => x.Key).ToList();
+            if (duplicateNames.Count != 0)
+            {
+                BH.Engine.Reflection.Compute.RecordError("Push failed: all tables need to have distinct names, regardless of letter casing.\n" +
+                                                        $"Following names are currently duplicate: {string.Join(", ", duplicateNames)}.");
+                return new List<object>();
+            }
```
I would move this up, before attempting to `check if the workbook exists and create it if not.`

alelom commented on Excel_Adapter/CRUD/Update/Update.cs:
```diff
@@ -0,0 +1,66 @@
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        private bool Update(IXLWorkbook workbook, Table table, ExcelPushConfig config)
```
It's fine if we do not override the default `Update` method, but I would still have this as `public`.

alelom commented on Excel_Adapter/CRUD/Update/UpdateWorkbookProperties.cs:
```diff
@@ -0,0 +1,66 @@
+        /**** Private Methods                           ****/
+        /***************************************************/
+
+        private void UpdateWorkbookProperties(IXLWorkbook workbook, WorkbookProperties properties)
```
Same comment as https://github.com/BHoM/Excel_Toolkit/pull/320/files#r757454175.

Further, I would follow the general naming conventions and have this method signature as:
```cs
public void Update(IXLWorkbook workbook, WorkbookProperties properties)
```

It is derivable form the parameters that we are updating a `Workbook`'s `properties` from the arguments, no need to have a long method name.

alelom commented on Excel_Engine/Create/CellContents.cs:
```diff
@@ -0,0 +1,84 @@
+        [Description("Created a BHoM wrapper for the given ClosedXML cell contents object.")]
+        [Input("xLCell", "ClosedXML cell contents object, on which the created BHoM wrapper is based.")]
+        [Output("cellContents", "BHoM wrapper for the input ClosedXML cell contents object.")]
+        public static CellContents CellContents(IXLCell xLCell)
```
Following the general convention, I would have this method as a `Convert` rather than a `Create`.

alelom commented on Excel_Engine/Create/CellContents.cs:
```diff
@@ -0,0 +1,84 @@
+        /**** Public Methods                    ****/
+        /*******************************************/
+
+        [Description("Created a BHoM wrapper for the given ClosedXML cell contents object.")]
```
```suggestion
        [Description("Converts a `ClosedXML` object into a `CellContent` BHoM object.")]
```

alelom reviewed (changes requested):
Just a few change requests from my point of view. 
I would also add the support for int&string column indexing as I don't think is too complex, but I'm happy to discuss that.

pawelbaran commented:
@BHoMBot check compliance
@BHoMBot check null-handling

pawelbaran commented:
Thank you @FraserGreenroyd @rolyhudson @alelom for the review. I think I addressed all comments, most of them rather trivial, only address/range requiring wider refactoring. Current solution includes:
- `CellAddress` with `string Column` and `int Row` in Excel convention (starting with "A" and 1)
- `CellRange` with `CellAddress From` and `CellAddress To`
- a set of `Create` methods allowing to create the above using strings in Excel format as well as programmatic indices (starting from 0)

I hope you will all like it 🙈

pawelbaran commented:
@BHoMBot check compliance
@BHoMBot check null-handling

pawelbaran commented:
@BHoMBot check core
@BHoMBot check installer
@BHoMBot check versioning
@BHoMBot check serialisation

rolyhudson commented on Excel_Engine/Query/ColumnName.cs:
```diff
@@ -0,0 +1,59 @@
+
+            string columnName = "";
+
+            while (columnIndex > 0)
```
An input of zero will return an empty name but should return A.
I think a special case is required for the 0 input?

pawelbaran replied:
Well spotted, you are right, the previous version did not work correctly. I have fixed the method and added a test script to the SP folder, just in case. Thanks!

rolyhudson reviewed (changes requested):
Two micro changes. A description update and handling a zero input on the ColumnName method. Otherwise this looks great.

rolyhudson reviewed (dismissed):
Nice work. Looking forward to leveraging this.

pawelbaran commented:
@BHoMBot check compliance
@BHoMBot check null-handling

pawelbaran commented:
@BHoMBot check core
@BHoMBot check installer
@BHoMBot check serialisation
@BHoMBot check versioning

FraserGreenroyd commented:
@BHoMBot check serialisation

pawelbaran commented:
@BHoMBot check serialisation

alelom commented on Excel_Adapter/ExcelAdapter.cs:
```diff
@@ -0,0 +1,67 @@
+        [Description("Specify Excel file and properties for data transfer.")]
+        [Input("fileSettings", "Input the file settings to get the file name and directory the Excel Adapter should use.")]
+        [Output("adapter", "Adapter to Excel.")]
+        public ExcelAdapter(BH.oM.Adapter.FileSettings fileSettings = null)
```
BHoM_Adapter's `FileSettings` is a bit controversial because it overlaps with File_Toolkit's `File`. I discussed in the past about deprecating it, but it can get incredibly complicated for reasons I will not expose here. Just leaving a comment for info, no action needed.

alelom commented on Excel_Engine/Create/CellAddress.cs:
```diff
@@ -0,0 +1,73 @@
+        [Input("columnIndex", "Index of the column to be used, starting from 0 equal to A in Excel.")]
+        [Input("rowIndex", "Index of the row to be used, starting from 0 equal to 1 in Excel.")]
+        [Output("address", "BHoM CellAddress object created based on the input indices.")]
+        public static CellAddress CellAddress(int columnIndex, int rowIndex)
+        {
+            string column = columnIndex.ColumnName();
+            if (string.IsNullOrWhiteSpace(column))
+                return null;
+
+            if (rowIndex < 0)
+            {
+                BH.Engine.Reflection.Compute.RecordError("Negative row index is not allowed.");
+                return null;
+            }
+
+            return new CellAddress { Column = column, Row = rowIndex + 1 };
+        }
```
I'd like to have this method as flexible as possible for columnIndex. Also, are we guarding against negative columnIndex? See suggestion below. I would even expand it more to cover `object rowIndex` too.

```suggestion
        public static CellAddress CellAddress(object columnIndex, int rowIndex)
        {
            Type columnIndexType = columnIndex.GetType();
            if (columnIndexType != typeof(string) && !columnIndexType.IsIntegralNumericType())
            {
                BH.Engine.Reflection.Compute.RecordError($"The {nameof(columnIndex)} input must either be a text indicating the Excel column name (e.g. 'AA') or an integer number indicating the column index.");
                return null;
            }
            
            string column = columnIndex.ToString();
            
            if (int.Parse(columnIndexType)< 0)
            {
                BH.Engine.Reflection.Compute.RecordError("Negative column index is not allowed.");
                return null;
            }
            
            if (rowIndex < 0)
            {
                BH.Engine.Reflection.Compute.RecordError("Negative row index is not allowed.");
                return null;
            }
            
            return new CellAddress { Column = column, Row = rowIndex + 1 };
```

alelom replied:
The reasons for this suggestion:
- flexibility from the UI
- flexibility when calling this method from code

alelom replied:
In other words, I would create a method `Query.IsValidCellIndexes(object columnIndex, object rowIndex)` that wraps up everything above except the CellAddress creation.

pawelbaran replied:
I have refactored the method and added `IsValidColumn`/`IsValidRow` checks to verify the inputs 👍

pawelbaran commented:
@BHoMBot check compliance
@BHoMBot check null-handling

pawelbaran commented:
@BHoMBot check null-handling

pawelbaran commented:
@BHoMBot check null-handling
@BHoMBot check compliance

pawelbaran commented:
@BHoMBot check core

pawelbaran commented:
@BHoMBot check installer
@BHoMBot check versioning
@BHoMBot check serialisation

alelom commented on Excel_Engine/Query/IsValid.cs:
```diff
@@ -0,0 +1,211 @@
+        public static bool IsValidColumn(this object column)
+        {
+            Type columnIndexType = column?.GetType();
+            if (columnIndexType != typeof(string) && !columnIndexType.IsIntegralNumericType())
```
Bug here: we need to allow for `double`s as input because Grasshopper sliders give `double` even for integer numbers.

We can correct this by doing the int.TryParse before.

Raising issue. Good to merge to get this in the Beta.

alelom replied:
![image](https://user-images.githubusercontent.com/6352844/144643627-e025c465-46c2-48bc-b376-a41cdc9cec64.png)

pawelbaran replied:
As discussed, let's merge this PR and fix this first thing next week, should be easy enough 👍

alelom replied:
Issue created: https://github.com/BHoM/Excel_Toolkit/issues/321

alelom commented on Excel_Engine/Query/IsValid.cs:
```diff
@@ -0,0 +1,211 @@
+        public static bool IsValidRow(this object row)
+        {
+            Type rowIndexType = row?.GetType();
+            if (rowIndexType != typeof(string) && !rowIndexType.IsIntegralNumericType())
```
Same bug as https://github.com/BHoM/Excel_Toolkit/pull/320/files#r762106787

alelom reviewed (approved):
With the following reserves, I am happy to approve in order to get this merged before Beta feature freeze.

- This bug: https://github.com/BHoM/Excel_Toolkit/pull/320/files#r762106787 needs to be fixed ASAP.  
  Relevant issue: https://github.com/BHoM/Excel_Toolkit/issues/321
- Minor comment to be considered in a future refactoring: https://github.com/BHoM/Excel_Toolkit/pull/320/files?authenticity_token=b0WcEPKISR1yK2xHqia%2BR6uYg46lfhrBNMA6lfMr3hmNxJ7D6iMujQ8eXrEHxpBU6%2B0K7fVKNOchsEoyompBHA%3D%3D&file-filters%5B%5D=.config&file-filters%5B%5D=.cs&file-filters%5B%5D=.sln#r761988397
